src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -203,10 +206,13 @@ public String url() {
      *  Note: to avoid deadlocks, when nesting locks we ALWAYS need to lock on the guard first, THEN the WorkflowRun.
      *  Synchronizing this helps ensure that fields are not mutated during a {@link #save()} operation, since that locks on the Run.
      */
-    private synchronized Object getMetadataGuard() {
-        if (metadataGuard == null) {
-            metadataGuard = new Object();
+    private Object getMetadataGuard() {
+        synchronized (this) {
+            if (metadataGuard == null) {
+                metadataGuard = new Object();
+            }
         }
+        assert !Thread.holdsLock(this) || Thread.holdsLock(metadataGuard): "Synchronizing on WorkflowRun before metadataGuard may cause deadlocks";
```

Comment:
I added an assertion for the condition given in the method's Javadoc to try to avoid any future issues. It causes a few tests to fail without the changes to `getListener` in this PR, mostly in `CpsPersistenceTest`.

Reply:
I'm pretty far out of my element here, but here goes. Aren't you doing the opposite of this (setting `synchronized (this)` on L211), and then initializing `metadataGuard` after doing so, just above? 

It's also unclear to me when `metaDataGuard` would be `null`, given it was initialized up on L195. This is the first time I've really dug into this so I'm probably just missing something obvious.

Reply:
That's a bit different. Lines 210-216 temporarily lock `this`, and then after that lock is released, the method returns `metadataGuard` for the caller to lock, so both locks are not held simultaneously. The concern here is only if we are holding both locks at the same time, which is what the assertion is checking.